### ISSUE TITLE
Props injected with getInitialProps appear to update on use of router.push with shallow routing - despite getInitialProps not being called again

### ISSUE THREAD

**Issue by benlester** [zeit/next.js#6747]
# Bug report

I have raised this behaviour on the next.js [Spectrum chat](https://spectrum.chat/next-js/general/props-from-getinitialprops-seem-to-be-updating-without-running-getinitialprops-again~854f1f59-e818-4e14-a18a-ef618114245d) to no avail. On talking to more people I believe this could be a bug. At the very least it appears to be terribly inconsistent behaviour.

I could be wrong, I'm still quite new to next.js - but I've tried everything I can think of.

## Describe the bug

I request a page, the request hits my back end express server, the query of the page is parsed and passed to my next.js app. 

The page is rendered, when rendering the page server side `getInitialProps` injects the query string of the page URL as props called `initialQuery`.

I have a component called `Search` on the page, the state of this component is initialised from the `initialQuery` props. Further changes to the state are pushed to the pages URL query string and stored locally in the state of `Search`.

I output the contents of `initialQuery` to the page, I expect that once the page is rendered by the server that `initialQuery` will not update - as all changes to the URL on the page are done using `{shallow: true}`, hence `getInitialProps` should not be called again.

I find that the output from `initialQuery` on the page changes when I change select/checkbox fields (manufacturer in MWE below), but **not** when I change `input` or `react-select` fields (low/high price, brand in MWE below).

As I said, I expect no change to `initialQuery` when using shallow routing on the page. But I find that inconsistently `initialQuery` is changed. I can't explain this behaviour. 

What's also very strange is the `getInitialProps` doesn't actually appear to be called again - despite the output it creates being updated on the page. I can't identify what updates the `initalQuery` props on the page.

## To Reproduce

Clone this really small MWE repo. I created here - [https://github.com/benlester/next-problem-example](https://github.com/benlester/next-problem-example)

To get going with it:

1. Clone the repo `git clone git@github.com:benlester/next-problem-example.git`
2. In frontend folder run `npm i`
3. In backend folder run `npm i`
4. Run the app. via the backend - `cd backend` -> `npm run start` (requires `nodemon`)

Once you've loaded the app. via `localhost:8080`:

1. Input a price into the low/high price - observe that the URL changes in a shallow manner **and** the rendered `initialQuery` on the page does not - this is expected behaviour when using shallow routing.
2. Select one of the manufacturers from the list - observe that the URL changes in a shallow manner **and** the rendered `initialQuery` on the page **does** change - this is not expected behaviour and seems inconsistent with the above.

## Expected behaviour

`initialQuery` in the MWE provided should not change unless the page is refreshed from the server. 

`initialQuery` should remain the same when using `router.push` with `{shallow: true}` option. 

Behaviour for `router.push` appears to be inconsistent depending on input type (this is the really strange bit)

## System information
 - OS: MacOS, Linux (tested on both)
 - Browser: Firefox, Chrome (tested on both)
 - Version of Next.js: 8.0.3

**Comment by codinronan** [comment]
Wow, interesting problem. This was a fun little puzzle to tackle.

TL;DR: This was your fault, but how you did it is really subtle. First things first, the problem is on this line:
https://github.com/benlester/next-problem-example/blob/master/frontend/components/search.js#L17

Let's consider what is actually happening there.

In `IndexPage.getInitialProps` you are doing the following:`

```
const initialQuery = parseQuery({ ...query })return { initialQuery }
```

Through Next's mechanisms, this data passes through `App.getInitialProps` to be returned as `pageProps.initialQuery`, which then becomes `props.initialQuery` in `IndexPage`, and which is then being passed wholesale through to your `Search` component - where your `Search` component then "makes a copy" of the object to avoid mutating it. All good, right?

You missed something.

In `lib/searchQuery.js` is this line:
```
searchQuery[field] = []
```

That same array is being passed down into `Search` - except you aren't copying it. You are copying `props.query` - which contains a reference to that array.

Then, in your `Search` component you do this when you change the checkbox:

```
const labelValue = this.state[label]
```
https://github.com/benlester/next-problem-example/blob/master/frontend/components/search.js#L57

You're mutating the array you "copied" in the constructor. You are mutating your state directly! THIS is why `initialQuery` appears to update on the home page - you mutated the `manufacturers` array referenced by `initialQuery` - it was never copied. You have the original reference that was created in `getInitialProps`!

One thing you should know is that _even though getInitialProps is not called on shallow pushes, the App component *still re-renders*_. It *must* in order to reflect the route change to consuming components. When you are mutating that array in memory, your re-render reflects the change. You are NOT mutating the initialQuery object when you add the price.

The solution to all this is simple. In your `Search` component constructor, you need a *deep copy* of the query:
```
this.state = { ...cloneDeep(props.query) }
```

Making that change, and the issue disappears and you no longer see `initialQuery` changing in the printout - as you would expect.

**Comment by benlester** [comment]
Once again, thank you so much for taking the time to help me with this issue. It means a lot. I'm always happy to be told I was incorrect when it means I learn something new.
